bug: Handle NoClassDefFoundError when loading YAML recipes with missing recipe dependencies

bryceatmoderne opened the pull request:
<!--
Thank you for taking the time to contribute to OpenRewrite!
Feel free to delete any sections that don't apply to your pull request.
-->

## What's changed?
Added logic to catch NoClassDefFoundError errors when loading YAML recipes with missing recipe dependencies

## What's your motivation?
NoClassDefFoundError errors were previously not handled compromising troubleshooting.

## Anything in particular you'd like reviewers to focus on?
Potential for regressed behaviour.

## Anyone you would like to review specifically?
@timtebeek 

## Have you considered any alternatives or workarounds?
No

## Any additional context
No

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [ ] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

timtebeek reviewed (approved):
Thanks for taking this on! I take it you've tested this particular scenario then as well?

timtebeek commented:
Had another look; when I modify the test case here to use an invalid class, it seems we're not entering the new catch block where I had expected that: 
https://github.com/openrewrite/rewrite/blob/03c458404073082275a752063887c188d213afbe/rewrite-core/src/test/java/org/openrewrite/config/YamlResourceLoaderTest.java#L303

![image](https://github.com/openrewrite/rewrite/assets/1027334/21062a27-d020-488e-927e-c50a8f89ce6d)

Also note that the new catch block is only added for Yaml resources that contain more than one document (or at least a `---` separator), whereas it might be quite common to find resources without such a block and handle missing classes there as well.

Perhaps we should still add a test if possible to ensure we're catching in the right place and not missing certain cases.

bryceatmoderne commented:
> Perhaps we should still add a test if possible to ensure we're catching in the right place and not missing certain cases.
I've added exception handling for the `String` `recipeData` case. I've also added unit testing as suggested. The tests validate the behaviour when a `NoClassDefFoundError` is thrown when attempting to instantiate a recipe, however it does so in a contrived manner that is not analogous to how the Moderne customer encountered it.
